go/extractor/cli/go-autobuilder/go-autobuilder.go

```diff
@@ -692,3 +695,161 @@ func main() {
+			"`environment.json` file specifying the maximum supported version of Go (" +
+			maxGoVersion + ")."
+		version = maxGoVersion
+		diagnostics.EmitNoGoModAndNoGoEnv(msg)
+	}
+
+	if !v.goInstallationFound && v.goDirectiveFound {
+		msg = "No version of Go installed. Writing an `environment.json` file specifying the " +
+			"version of Go found in the `go.mod` file (" + v.goModVersion + ")."
+		version = v.goModVersion
+		diagnostics.EmitNoGoEnv(msg)
+	}
+
+	if v.goInstallationFound && !v.goDirectiveFound {
+		msg = "No `go.mod` file found. Version " + v.goEnvVersion + " installed in the environment."
+		version = ""
+		diagnostics.EmitNoGoMod(msg)
+	}
+
+	return msg, version
+}
+
+func compareVersions(v versionInfo) (msg, version string) {
+	if semver.Compare("v"+v.goModVersion, "v"+v.goEnvVersion) > 0 {
+		msg = "The version of Go installed in the environment (" + v.goEnvVersion +
+			") is lower than the version found in the `go.mod` file (" + v.goModVersion +
+			").\nWriting an `environment.json` file specifying the version of Go from the " +
+			"`go.mod` file (" + v.goModVersion + ")."
+		version = v.goModVersion
+		diagnostics.EmitVersionGoModHigherVersionEnvironment(msg)
+	} else {
+		msg = "The version of Go installed in the environment (" + v.goEnvVersion +
+			") is high enough for the version found in the `go.mod` file (" + v.goModVersion + ")."
+		version = ""
+		diagnostics.EmitVersionGoModNotHigherVersionEnvironment(msg)
+	}
+
+	return msg, version
+}
+
+func getVersionToInstall(v versionInfo) (msg, version string) {
+	msg, version = checkForUnsupportedVersions(v)
+	if msg != "" {
+		return msg, version
+	}
+
+	msg, version = checkForVersionsNotFound(v)
+	if msg != "" {
+		return msg, version
+	}
+
+	msg, version = compareVersions(v)
+	return msg, version
+}
+
+func writeEnvironmentFile(version string) {
+	var content string
+	if version == "" {
+		content = `{ "include": [] }`
+	} else {
+		content = `{ "include": [ { "go": { "version": "` + version + `" } } ] }`
+	}
+
+	targetFile, err := os.Create("environment.json")
+	if err != nil {
+		log.Println("Failed to create environment.json: ")
+		log.Println(err)
+		return
+	}
+	defer func() {
+		if err := targetFile.Close(); err != nil {
+			log.Println("Failed to close environment.json:")
+			log.Println(err)
+		}
+	}()
+
+	_, err = targetFile.WriteString(content)
+	if err != nil {
+		log.Println("Failed to write to environment.json: ")
+		log.Println(err)
+	}
+}
+
+type versionInfo struct {
+	goModVersion        string
+	goDirectiveFound    bool
+	goEnvVersion        string
+	goInstallationFound bool
+}
+
+func (v versionInfo) String() string {
+	return fmt.Sprintf("go.mod version: %s, go.mod directive found: %t, go env version: %s, go installation found: %t", v.goModVersion, v.goDirectiveFound, v.goEnvVersion, v.goInstallationFound)
+}
+
+func isGoInstalled() bool {
+	_, err := exec.LookPath("go")
+	return err == nil
+}
```

Comment:
It's a well-known program so I'm happy to take that risk. We will fall over very soon if it isn't the go compiler because we'll try to parse the output of `go version`.